### PR TITLE
auth/clientcredentials: add schema-based scope enforcement interceptor

### PR DESCRIPTION
**Pull request by bobheadxi**
Adds an interceptor, `clientcredentials.Interceptor`, that enforces required SAMS scopes based on a proto schema extension. This will reduce boilerplate in RPC implementations and reduce the chance of mistakes/out-of-sync problems between a schema and the implementation

Usage example:

```proto
extend google.protobuf.MethodOptions {
	// The SAMS scopes required to use this RPC.
	//
	// The range 50000-99999 is reserved for internal use within individual organizations
	// so you can use numbers in this range freely for in-house applications.
	repeated string sams_required_scopes = 50001;
}
```

Allows you to set required scopes as method options:

```proto
rpc GetUserRoles(GetUserRolesRequest) returns (GetUserRolesResponse) {
	option (sams_required_scopes) = "sams::user.roles::read";
};
```

This generates `E_SamsRequiredScopes` that can be used to point to where we can extract `sams_required_scopes`.

## Test plan

Unit tests

### CODE REVIEW

File: auth/clientcredentials/clientcredentials.go [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+	// Check for our required scope.
+	for _, required := range requiredScopes {
+		if !result.Scopes.Match(required) {
+			// Record detailed error in span and logs
+			err = errors.Newf("got scopes %+v, required: %+v", result.Scopes, requiredScopes)
+			span.SetAttributes(attribute.String("full_error", err.Error()))
+			logger.Error("attempt to authenticate using SAMS token without required scope",
+				log.Error(err))
+			// Return an opaque error
+			return info, connect.NewError(connect.CodePermissionDenied, errors.New("insufficient scope"))
```
Comment (unknwon) [inline auth/clientcredentials/clientcredentials.go:168]: I think:
1. Logging service-side errors here is non-actionable, after all this is client's problem to have insufficient scope.
2. The error returned to the client does not need to be opaque (no security concerns), and being explicit greatly help the client debug scope issues.

File: auth/clientcredentials/clientcredentials.go [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+func extractBearerContents(h http.Header) (string, error) {
+	authHeader := h.Get("Authorization")
+	if authHeader == "" {
+		return "", errors.New("no token provided in Authorization header")
+	}
+	typ := strings.SplitN(authHeader, " ", 2)
+	if len(typ) != 2 {
+		return "", errors.New("token type missing in Authorization header")
+	}
+	if strings.ToLower(typ[0]) != "bearer" {
```
Comment (unknwon) [inline auth/clientcredentials/clientcredentials.go:206]:
```suggestion
	if strings.EqualFold(typ[0], "bearer") {
```

can avoid string allocation.

File: auth/clientcredentials/clientcredentials.go (hunk 1 above)
Comment (bobheadxi) [reply auth/clientcredentials/clientcredentials.go]: Removed the log, and made the error non-opaque 👍

File: auth/clientcredentials/clientcredentials_test.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+		},
+		wantError: autogold.Expect("unauthenticated: invalid authorization header: invalid token type Bearer in Authorization header"), // should not error!
+		wantLogs:  autogold.Expect([]string{}),
+	}, {
+		name: "wrong scope",
+		token: &sams.IntrospectTokenResponse{
+			Active: true,
+			Scopes: scopes.Scopes{"not-a-scope"},
+		},
+		wantError: autogold.Expect("unauthenticated: invalid authorization header: invalid token type Bearer in Authorization header"),
```
Comment (unknwon) [inline auth/clientcredentials/clientcredentials_test.go:65]: Are we sure we want errors to be asserted here be `invalid authorization header`? 😂 shouldn't them be like something `missing required scope: xxx`?

Comment (bobheadxi) [reply auth/clientcredentials/clientcredentials_test.go:65]: oof good catch, I think I broke something

Comment (bobheadxi) [reply auth/clientcredentials/clientcredentials_test.go:65]: I think https://github.com/sourcegraph/sourcegraph-accounts-sdk-go/pull/61/commits/e0d195ce3da572f65c7429c3f1392ab7602560db broke it (the equalfold change)

Comment (bobheadxi) [reply auth/clientcredentials/clientcredentials_test.go:65]: condition was inverted hahaha

Comment (bobheadxi) [reply auth/clientcredentials/clientcredentials_test.go]: updated!

**Review by unknwon — APPROVED** [review]

File: auth/clientcredentials/clientcredentials_test.go (hunk 3 above)
Comment (unknwon) [reply auth/clientcredentials/clientcredentials_test.go]: 😂

**Pull request merged**